aruha-399 compatibility mode

rcillo opened the pull request:
Documentation for the three compatibility modes. It focuses on the
guarantees provided by each mode so that users know what to expected
from event types using each one of them.

Commit:
aruha-399 compatibility mode

Documentation for the three compatibility modes. It focuses on the
guarantees provided by each mode so that users know what to expected
from event types using each one of them.

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
 
+### Compatibility modes
+
+Event type schemas evolve through time and different data integration
```
`time and` -> `time, and`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+patterns require different guarantees when it comes to what changes
+should or should not be allowed.
+
+To accomodate these different needs Nakadi provides three distinct
```
`accomodate` -> `accommodate`

lmontrieux replied:
You can also remove `distinct` (only if you want to), the sentence still works without it

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+The complete set of json-schema is supported, including
+additionalProperties. Consumers willing to read events in a safe
+manner should ignore fields that are not declared in the
+schema. Making assumptions about non declared properties might lead to
```
`non declared` -> `undeclared`?

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+
+This is the most strict mode. It guarantees that events published
+under this mode can be safely read using both newer and older schema
+versions. E.g. it's both forward compatible and backwards
```
I think you mean i.e., not e.g.?

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+
+1. Clients using old schemas can safely read events validated by newer
+   schemas.
+2. Clients using the newest schema can safely read events validated by
```
`newest` -> `newer`, if I'm not mistaken

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+2. Clients using the newest schema can safely read events validated by
+   older schemas.
+
+In order to achieve that, only a subset of json schema will be allowed
```
Earlier you wrote `json-schema`, not `json schema`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+   older schemas.
+
+In order to achieve that, only a subset of json schema will be allowed
+to make sure no accidental incompatibilities or ambiguities will be
```
`will be` -> `are`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+
+In order to achieve that, only a subset of json schema will be allowed
+to make sure no accidental incompatibilities or ambiguities will be
+introduced. Only compatible changes aregoing to be allowed when
```
`aregoing` -> `are going`

lmontrieux replied:
In fact, `aregoing to be` -> `are`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+In order to achieve that, only a subset of json schema will be allowed
+to make sure no accidental incompatibilities or ambiguities will be
+introduced. Only compatible changes aregoing to be allowed when
+changing the schema. This mode will be required to have data stored
```
`will be` -> `is` (unless we want to change this again when it becomes true)

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+to make sure no accidental incompatibilities or ambiguities will be
+introduced. Only compatible changes aregoing to be allowed when
+changing the schema. This mode will be required to have data stored
+and processed by the data lake.
```
`to have data stored and processed by the data lake` -> `for the data lake to store and process data` (removing passive form)

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+
+Schemas can be changed arbitrarily. Nakadi will automatically manage
+the version field so that if an incompatible change happens (MAJOR
+version bump) consumers are going to be aware of it.
```
`are going to be` -> `are`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,69 @@ The custom schema for an event type can be as simple as
+When using **compatible** mode, validation of events accept only
+properties that are explicitly stated in the schema. It means that
+events containing attributes that cannot be mapped to a schema
+property are going to be rejected.
```
`are going to be` -> `will be`

Commit:
aruha-399 provide more context on modes

It's important to explain the behaviour, the use case and the
limitations of each mode.

rcillo commented:
@lmontrieux thank you for the previous review. I applied the changes when appropriate but changed a lot the text. It wasn't properly providing all the context, behaviour, limitations and use case for each mode.

I hope it's clearer now.

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+### Compatibility modes
+
+Compatibility modes are used to control schema changes. Each mode
+solves a specific problem and thus present different constraints. It
```
present -> presents

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+
+Compatibility modes are used to control schema changes. Each mode
+solves a specific problem and thus present different constraints. It
+ranges from being very constraining, allowing just some minor changes
```
comma (,) after 'changes'

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+to allowing everything.
+
+By default, Nakadi uses the forward mode, which is further explained
+bellow, together with the other two modes: compatible and none.
```
`bellow` -> `below`

lmontrieux replied:
`the other two modes` -> `the two other modes` (I think)

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+This mode has been designed to allow event type owners to change their
+schemas without breaking existing consumers.
+
+Its called forward because using older schemas it's safe to read
```
`Its` -> `It's` (or better, `It is`)

lmontrieux replied:
comma after `older schemas`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+
+Its called forward because using older schemas it's safe to read
+events generated by newer schemas. On the other hand, it's not
+backward, in the sense that trying to read an older event using a
```
`backward` -> `backward compatible`, if I'm not mistaken?

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+backward, in the sense that trying to read an older event using a
+newer schema is not safe.
+
+So if the consumer has been developed with schema version `1.5.0` in
```
`So` -> `For example,`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+
+So if the consumer has been developed with schema version `1.5.0` in
+mind, reading events from `1.5.0` on is safe, for example, `1.6.0`. But
+reading event generated by older schemas, like `1.4.0` is not safe.
```
comma after `1.4.0`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+ignore event attributes not declared in the event type schema.
+
+This mode is default because it solves the most common requirement:
+guarantee stability of deployed consumer applications when schema
```
`guarantee stability` -> `guarantee the stability`
`consumer` -> `consumers`
`when schema` -> `when the schema`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+2. Clients using newer schema can safely read events validated by
+   older schemas.
+
+It's useful for long term data storage and analysis. It's guarantees
```
`It's guarantees` -> `It guarantees`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+   older schemas.
+
+It's useful for long term data storage and analysis. It's guarantees
+the highest quality data for business intelegence.
```
`intelegence` -> `intelligence`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+It's useful for long term data storage and analysis. It's guarantees
+the highest quality data for business intelegence.
+
+Only a subset of json schema is allowed to make sure no accidental
```
`allowed to` -> `allowed, to`
`sure no` -> `sure that no`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+the highest quality data for business intelegence.
+
+Only a subset of json schema is allowed to make sure no accidental
+incompatibilities or ambiguities will be introduced. This mode is
```
`will be` -> `are`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+
+Only a subset of json schema is allowed to make sure no accidental
+incompatibilities or ambiguities will be introduced. This mode is
+required to have data stored and processed by the data lake.
```
`to have data` -> `for data to be`
OR
change the whole line to `required for the data lake to store and process the data.`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+This is the most strict mode. It's both forward compatible and
+backward compatible. In practical terms:
+
+1. Consuemrs using older schemas can safely read events validated by
```
`Consuemrs` -> `Consumers`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+Supported changes:
+
+1. changes to meta attributes: titles, descriptions and examples.
+2. addition of new attributes.
```
both optional and required, right? Maybe this should be explicit, as in the Compatible you say that only new optional attributes are allowed

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+3. not
+4. patternProperties
+
+Removing the support to these features is necessary to avoid the
```
`support to` -> `support for`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+
+Removing the support to these features is necessary to avoid the
+introduction of incompatible changes. Since additionalProperties is
+not supported, it's value is defaulted to false: all events with
```
`it's value is defaulted to false` -> `its value is false by default`

lmontrieux commented on docs/using/event-types.md:
```diff
@@ -44,6 +44,83 @@ The custom schema for an event type can be as simple as
+Removing the support to these features is necessary to avoid the
+introduction of incompatible changes. Since additionalProperties is
+not supported, it's value is defaulted to false: all events with
+undeclared attributes rejected.
```
`attributes rejected` -> `attributes will be rejected`

lmontrieux reviewed (commented):
Look very clear and detailed to me. I mostly made comments about grammar and typos.

Commit:
aruha-399 fix grammar

rcillo commented:
👍 made all the corrections. I realised my grammar checker was disabled 😅

lmontrieux commented:
👍
